drivers/counter/counter_native_posix.c

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2020, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device.h>
+#include <drivers/counter.h>
+#include <soc.h>
+#include <hw_counter.h>
+
+#define DT_COUNTER_LABEL counter0
+#define DRIVER_CONFIG_INFO_FLAGS (COUNTER_CONFIG_INFO_COUNT_UP)
+#define DRIVER_CONFIG_INFO_CHANNELS 1
+#define COUNTER_NATIVE_POSIX_IRQ_FLAGS (0)
+#define COUNTER_NATIVE_POSIX_IRQ_PRIORITY (2)
+
+#define COUNTER_PERIOD (USEC_PER_SEC / CONFIG_COUNTER_NATIVE_POSIX_FREQUENCY)
+#define TOP_VALUE (UINT_MAX)
+
+static struct counter_alarm_cfg pending_alarm;
+static bool is_alarm_pending;
+static struct device *device;
+
+static void counter_isr(void *arg)
+{
+	ARG_UNUSED(arg);
+	uint32_t current_value = hw_counter_get_value();
+
+	if (is_alarm_pending) {
+		is_alarm_pending = false;
+		pending_alarm.callback(device, 0, current_value,
+				       pending_alarm.user_data);
+	}
+}
+
+static int ctr_init(struct device *dev)
+{
+	device = dev;
+	is_alarm_pending = false;
+
+	IRQ_CONNECT(COUNTER_EVENT_IRQ, COUNTER_NATIVE_POSIX_IRQ_PRIORITY,
+		    counter_isr, NULL, COUNTER_NATIVE_POSIX_IRQ_FLAGS);
+	hw_counter_set_period(COUNTER_PERIOD);
+	hw_counter_set_target(TOP_VALUE);
+
+	return 0;
+}
+
+static int ctr_start(struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	hw_counter_start();
+	return 0;
+}
+
+static int ctr_stop(struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	hw_counter_stop();
+	return 0;
+}
+
+static int ctr_get_value(struct device *dev, uint32_t *ticks)
+{
+	ARG_UNUSED(dev);
+
+	*ticks = hw_counter_get_value();
+	return 0;
+}
+
+static uint32_t ctr_get_pending_int(struct device *dev)
+{
+	ARG_UNUSED(dev);
+	return 0;
+}
+
+static int ctr_set_top_value(struct device *dev,
+			     const struct counter_top_cfg *cfg)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(cfg);
+
+	posix_print_warning("%s not supported\n", __func__);
+	return -ENOTSUP;
+}
+
+static uint32_t ctr_get_top_value(struct device *dev)
+{
+	return TOP_VALUE;
+}
+
+static uint32_t ctr_get_max_relative_alarm(struct device *dev)
+{
+	return TOP_VALUE;
+}
+
+static int ctr_set_alarm(struct device *dev, uint8_t chan_id,
+			 const struct counter_alarm_cfg *alarm_cfg)
+{
+	ARG_UNUSED(dev);
+
+	if (chan_id >= DRIVER_CONFIG_INFO_CHANNELS) {
+		posix_print_warning("channel %u is not supported\n", chan_id);
+		return -ENOTSUP;
+	}
+
```

Comment:
just noticed that it is not documented in the API but other implementations returns -EBUSY if channel is already configured and did not yet expired.

Reply:
So, would it conventionally be better to force user to cancel pending alarm (by returning -EBUSY) before setting it up again on a given channel? Did I understand it correctly?